File: apps/api/src/application/products/update-product.ts
import { normalizeCategoryName, normalizeSku } from '@/domain/product/product'
import { NotFoundError } from '@/domain/shared/domain-error'

import type { CategoryRepository } from '@/application/ports/category-repository'
import type { ProductChanges, ProductRepository } from '@/application/ports/product-repository'
import type { Product } from '@/domain/product/product'
import type { UpdateProduct as UpdateProductInput } from '@ecommerce/shared'

export class UpdateProduct {
  constructor(
    private readonly products: ProductRepository,
    private readonly categories: CategoryRepository,
  ) {}

  async execute(id: string, input: UpdateProductInput): Promise<Product> {
    const changes: ProductChanges = {
      ...(input.sku !== undefined && { sku: normalizeSku(input.sku) }),
      ...(input.name !== undefined && { name: input.name }),
      ...(input.description !== undefined && { description: input.description ?? null }),
      ...(input.price !== undefined && { price: input.price }),
      ...(input.stock !== undefined && { stock: input.stock }),
      ...(input.weightKg !== undefined && { weightKg: input.weightKg ?? null }),
      ...(await this.categoryChange(input.category)),
    }
    const product = await this.products.update(id, changes)
    if (!product) {
      throw new NotFoundError('product', id)
    }
    return product
  }

  private async categoryChange(
    category: string | null | undefined,
  ): Promise<Pick<ProductChanges, 'categoryId'>> {
    if (category === undefined) {
      return {}
    }
    if (category === null) {
      return { categoryId: null }
    }
    const found = await this.categories.findOrCreate(normalizeCategoryName(category))
    return { categoryId: found.id }
  }
}

File: apps/api/src/ecommerce_api/application/products/update_product.py
from ecommerce_api.application.ports.category_repository import CategoryRepository
from ecommerce_api.application.ports.product_repository import ProductChanges, ProductRepository
from ecommerce_api.application.schemas.product import UpdateProductInput
from ecommerce_api.domain.errors import NotFoundError
from ecommerce_api.domain.product import Product, normalize_category_name, normalize_sku


class UpdateProduct:
    def __init__(self, products: ProductRepository, categories: CategoryRepository) -> None:
        self._products = products
        self._categories = categories

    async def execute(self, id: str, data: UpdateProductInput) -> Product:
        changes = ProductChanges()
        if data.sku is not None:
            changes['sku'] = normalize_sku(data.sku)
        if data.name is not None:
            changes['name'] = data.name
        if data.has('description'):
            changes['description'] = data.description
        if data.price is not None:
            changes['price'] = data.price
        if data.stock is not None:
            changes['stock'] = data.stock
        if data.has('weight_kg'):
            changes['weight_kg'] = data.weight_kg
        if data.has('category'):
            changes['category_id'] = await self._category_id(data.category)
        product = await self._products.update(id, changes)
        if product is None:
            raise NotFoundError('product', id)
        return product

    async def _category_id(self, category: str | None) -> str | None:
        if category is None:
            return None
        return (await self._categories.find_or_create(normalize_category_name(category))).id
